curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -1196,22 +1206,261 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
+        pagePos = blockPos % pageSize;
+        while (blockLen > 0) {
+            if (pagePos + blockLen > pageSize) {
+                m = pageSize - pagePos;
+            } else {
+                m = blockLen;
+            }
+
+            PageData* pageData = new PageData();
+            pageData->data = new char[pageSize];
+            if (!pageData->data) {
+                LOG(ERROR) << "new page failed.";
+                return;
+            }
+            memset(pageData->data, 0, pageSize);
+            memcpy(pageData->data + pagePos, data + dataOffset, m);
+            if (pagePos + m < pageSize) {
+                tailZeroLen = pageSize - pagePos - m;
+            }
+            pageData->index = pageIndex;
+            assert(pdMap.count(pageIndex) == 0);
+            pdMap.emplace(pageIndex, pageData);
+            pageIndex++;
+            blockLen -= m;
+            dataOffset += m;
+            pagePos = (pagePos + m) % pageSize;
+        }
+
+        blockIndex++;
+        len -= n;
+        blockPos = (blockPos + n) % blockSize;
+    }
+    actualLen_ = headZeroLen + len_ + tailZeroLen;
+    assert((actualLen_ % pageSize) == 0);
+    assert((actualChunkPos_ % pageSize) == 0);
+    createTime_ = ::curve::common::TimeUtility::GetTimeofDaySec();
+}
+
+void DataCache::CopyDataToDataCache(uint64_t DataCachePos, uint64_t len,
+                                    const char* data) {
+    uint64_t blockSize = s3ClientAdaptor_->GetBlockSize();
+    uint32_t pageSize = s3ClientAdaptor_->GetPageSize();
+    uint64_t pos = chunkPos_ + DataCachePos;
+    uint64_t blockIndex = pos / blockSize;
+    uint64_t blockPos = pos % blockSize;
+    uint64_t pageIndex, pagePos;
+    uint64_t n, blockLen, m;
+    uint64_t dataOffset = 0;
+    uint64_t addLen = 0;
+
+    VLOG(9) << "CopyDataToDataCache() DataCachePos:" << DataCachePos
+            << ", len:" << len << ", chunkPos_:" << chunkPos_
+            << ", len_:" << len_;
+    if (DataCachePos + len > len_) {
+        len_ = DataCachePos + len;
+    }
+    while (len > 0) {
+        if (blockPos + len > blockSize) {
+            n = blockSize - blockPos;
+        } else {
+            n = len;
+        }
+        blockLen = n;
+        PageDataMap& pdMap = dataMap_[blockIndex];
+        PageData* pageData;
+        pageIndex = blockPos / pageSize;
+        pagePos = blockPos % pageSize;
+        while (blockLen > 0) {
+            if (pagePos + blockLen > pageSize) {
+                m = pageSize - pagePos;
+            } else {
+                m = blockLen;
+            }
+            if (pdMap.count(pageIndex)) {
+                pageData = pdMap[pageIndex];
+            } else {
+                pageData = new PageData();
+                pageData->data = new char[pageSize];
+
+                if (!pageData->data) {
+                    LOG(ERROR) << "new page failed.";
+                    return;
+                }
+                memset(pageData->data, 0, pageSize);
```

Comment:
if user just writes 4k data within current page, will you upload a 64k object to s3?